### PR TITLE
Fixes #3924

### LINKED ISSUE #3924: RGD default scoring function does not always work as expected
**Describe the bug**

In RGD `testRGroupDecomp.cpp` tests `testRingMatching3` and `testGeminalRGroups` fail unless the fingerprint scoring function is used.  In each case a obviously inferior decomposition is created when the default scoring function is used

**To Reproduce**

In both test functions comment out the line  `params.scoreMethod = FingerprintVariance;`

**Expected behavior**

The test passes using the default scoring function

**Screenshots**

Expected behavior

Capture2

Capture2

Actual behavior

Capture2

Capture2

**Configuration (please complete the following information):**
 - RDKit version:  master as of 3/13/21
 - OS: Ubuntu 18:04

### PR DESCRIPTION
**Pull request by bp-kelley**
The scoring function was not taking into account empty rgroups which biased certain arrangements of rgroups that didn't have an rgroup at every position.

@ptosco @jones-gareth  @greglandrum This resolves #3924 and mercifully doesn't fail any of the existing regression tests!

### CODE REVIEW

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp [hunk 1]
```diff
@@ -76,16 +90,22 @@ double matchScore(const std::vector<size_t> &permutation,
...
                   << linkerMatchSet[rg->second->attachments] << std::endl;
 #endif
         ++i;
       }
     }
-
+    
     double tempScore = 0.;
-    for (const auto &matchSet : matchSetVect) {
+    const std::string no_rgroup;
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:102]: Why is this uninitialized?

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp [hunk 2]
```diff
@@ -38,20 +38,34 @@ double matchScore(const std::vector<size_t> &permutation,
...
+  }
+  // for each label (r-group)
   for(auto l : labels ) {
 #ifdef DEBUG
     std::cerr << "Label: " << l << std::endl;
 #endif
     std::vector<std::map<std::string, unsigned int>> matchSetVect;
     std::map<std::set<int>, size_t> linkerMatchSet;
 
+    int num_rgroups_for_label = 0;
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:61]: It looks like this is only being used inside the `#ifdef DEBUG`. Maybe worth either removing or doing:
```suggestion
#ifdef DEBUG
    int num_rgroups_for_label = 0;
#endif
```

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp [hunk 3]
```diff
@@ -38,20 +38,34 @@ double matchScore(const std::vector<size_t> &permutation,
...
     for (size_t m = 0; m < permutation.size(); ++m) {  // for each molecule
 
       auto rg = matches[m][permutation[m]].rgroups.find(l);
       if (rg == matches[m][permutation[m]].rgroups.end()) {
         continue;
       }
+      num_rgroups_for_label++;
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:68]: It looks like this is only being used inside the `#ifdef DEBUG`. Maybe worth either removing or doing:
```suggestion
#ifdef DEBUG
      num_rgroups_for_label++;
#endif
```

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp [hunk 4]
```diff
@@ -38,20 +38,34 @@ double matchScore(const std::vector<size_t> &permutation,
...
+  // What is the largest rgroup count at any label
+  int N = 0;
+  std::map<int, int> num_rgroups;
+  for (size_t m = 0; m < permutation.size(); ++m) {  // for each molecule
+  for (auto l : matches[m][permutation[m]].rgroups) {
+    int count = ++num_rgroups[l.first];
+    if (N < count) {
+        N = count;
+      }
+    }
```
Comment (greglandrum) [inline Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:51]:
```suggestion
   N = std::max(N,count);
```

**Review by greglandrum — CHANGES_REQUESTED** [review]
small changes

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp (hunk 1 above)
Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:102]: It is initialized, It's just using the default constructor.

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp (hunk 3 above)
Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:68]: It is being used in the matchSet loop.

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:102]: ok, let me rephrase: "Why are you using an empty key?"

Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:102]: I'll make it more explicit if you like.

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp (hunk 3 above)
Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:68]: sorry, missed that.

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp (hunk 1 above)
Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:102]: I'm not saying that it's wrong, it's just unusual enough that it would be worth adding a comment.

Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:99]: Happy to change it, but here are the default initialization rules
```
> What is default initialization?

a. For a POD class type or a built-in POD type like int, initialise to an indeterminate value (do nothing).
b. For a non-POD class type, call a default constructor which provides the initialised value.
c. For an array type, default initialise every element of the array. 
```

Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:99]: I believe this was done partly to solve the "vexing parse" issue

```
std::string Foo();
```

Is that a variable or a function definition?

Comment (greglandrum) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:99]: My suggestion was that it's worth adding a comment to the code to explain that you are intentionally using an empty string as a key in the map.

Comment (bp-kelley) [reply Code/GraphMol/RGroupDecomposition/RGroupScore.cpp]: I did a small refactor and changed the name to EMPTY_RGROUP

**Review by greglandrum — APPROVED** [review]

**Review by greglandrum — CHANGES_REQUESTED** [review]
I'm changing my answer on this one.
It should include something about the two tests called out in #3924

**Review by greglandrum — APPROVED** [review]

**Comment by greglandrum** [comment]
@jones-gareth : If you are ok with these set of changes I'll go ahead and merge them.

File: Code/GraphMol/RGroupDecomposition/testRGroupInternals.cpp [hunk 5]
```diff
@@ -168,7 +168,7 @@ void testRingMatching3Score() {
   // expect test1 to have better score than test2 since all halogens are on R1
   // but this is not the case
```
Comment (jones-gareth) [inline Code/GraphMol/RGroupDecomposition/testRGroupInternals.cpp:169]: Now it's fixed we can remove this line

File: Code/GraphMol/RGroupDecomposition/RGroupScore.cpp [hunk 6]
```diff
@@ -63,7 +74,7 @@ double matchScore(const std::vector<size_t> &permutation,
 #endif
       }
 #ifdef DEBUG
-      std::cerr << " " << rg->second->combinedMol->getNumAtoms(false)
+        std::cerr << l << " rgroup count" << num_rgroups_for_label << " num atoms" << rg->second->combinedMol->getNumAtoms(false)
                 // looks like code has been edited round this define
```
Comment (jones-gareth) [inline Code/GraphMol/RGroupDecomposition/RGroupScore.cpp:78]: I put in this comment as score is not defined.  If you don't need it perhaps delete both rows

**Comment by jones-gareth** [comment]
> @jones-gareth : If you are ok with these set of changes I'll go ahead and merge them.

I've added a couple of comments.

Also in `testRGroupDecomp.cpp` both `testMutipleCoreRelabellingIssues` and `testUnprocessedMapping` are using the `fingerprintVariance` score.  Probably worth checking that the default scoring function is not working for those tests- perhaps test both scoring functions as you've done with `testRingMatching3` (which is a nice feature).

**Comment by bp-kelley** [comment]
I can check these two, but probably not until tomorrow.

**Comment by greglandrum** [comment]
I removed the two comments suggested by @jones-gareth . Will merge after the CI builds pass
